src/hacbs/components/ApplicationDetails/tabs/overview/visualization/hooks/useAppWorkflowData.ts

```diff
@@ -1,129 +1,582 @@
+            new Date(a.metadata.creationTimestamp).getTime(),
+        )?.[0];
+      if (latestBuild) {
+        acc.push(latestBuild);
+      }
+      return acc;
+    }, [] as PipelineRunKind[]);
+  }, [buildPipelines, buildPipelinesLoaded, components, componentsLoaded]);
+
+  const buildNodes: WorkflowNodeModel<WorkflowNodeModelData>[] = React.useMemo(() => {
+    const nodes =
+      components.length && latestBuilds.length
+        ? components.map((component) => getBuildNodeForComponent(component, latestBuilds))
+        : [emptyPipelineNode('no-builds', 'Builds', WorkflowNodeType.BUILD, componentTasks)];
+    const maxWidth = nodes.reduce((max, node) => Math.max(max, node.width), 0);
+    nodes.forEach((n) => (n.width = maxWidth));
+    return nodes;
+  }, [componentTasks, components, latestBuilds]);
+
+  const buildGroup = React.useMemo(
+    () =>
+      allResourcesLoaded
+        ? groupToPipelineNode(
+            'builds',
+            'Builds',
+            WorkflowNodeType.BUILD,
+            componentTasks,
+            expanded,
+            expanded ? buildNodes?.map((c) => c.id) : undefined,
+            buildNodes,
+            latestBuilds,
+            worstWorkflowStatus(buildNodes),
+          )
+        : undefined,
+    [allResourcesLoaded, buildNodes, componentTasks, latestBuilds, expanded],
+  );
+
+  const buildTasks = React.useMemo(
+    () => (expanded && buildNodes?.length ? buildNodes.map((c) => c.id) : [buildGroup?.id ?? '']),
+    [buildGroup?.id, buildNodes, expanded],
+  );
 
-  componentIntegrationTests = integrationTestScenario?.filter((test) => {
+  const componentIntegrationTests = integrationTestScenario?.filter((test) => {
     const contexts = test?.spec?.contexts;
     return contexts?.some((c) => c.name === 'component') ?? false;
   });
 
-  applicationIntegrationTests = integrationTestScenario?.filter((test) => {
+  const componentTestNodes = React.useMemo(() => {
+    if (!allResourcesLoaded) {
+      return [];
+    }
+    const nodes = componentIntegrationTests?.length
+      ? componentIntegrationTests.map((test) =>
+          resourceToPipelineNode(test, WorkflowNodeType.COMPONENT_TEST, buildTasks),
+        )
+      : [
+          emptyPipelineNode(
+            'no-component-tests',
+            'Component tests',
+            WorkflowNodeType.COMPONENT_TEST,
+            buildTasks,
+          ),
+        ];
+    const maxWidth = nodes.reduce((max, node) => Math.max(max, node.width), 0);
+    nodes.forEach((n) => (n.width = maxWidth));
+    return nodes;
+  }, [allResourcesLoaded, buildTasks, componentIntegrationTests]);
+
+  const componentIntegrationTestTasks = React.useMemo(
+    () => (expanded ? componentTestNodes.map((c) => c.id) : []),
+    [componentTestNodes, expanded],
+  );
+
+  const applicationIntegrationTests = integrationTestScenario?.filter((test) => {
     const contexts = test?.spec?.contexts;
     return contexts?.some((c) => c.name === 'application') ?? false;
   });
```

Comment:
You need to use `useMemo` here since `applicationIntegrationTests` is used as a memo dependency. Otherwise the dependency check will always fail.